Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,709 @@
+﻿#if UNITY_EDITOR
+using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR.Utilities;
+using UnityEngine;
+
+namespace UnityEditor.Experimental.EditorVR.Modules
+{
+	[MainMenuItem("Snapping", "Settings", "Select snapping modes")]
+	sealed class SnappingModule : MonoBehaviour, IUsesViewerScale, ISettingsMenuProvider
+	{
+		const float k_GroundSnappingMaxRayLength = 25f;
+		const float k_SurfaceSnappingMaxRayLength = 100f;
+
+		const float k_GroundHeight = 0f;
+
+		const float k_ManipulatorGroundSnapMin = 0.05f;
+		const float k_ManipulatorGroundSnapMax = 0.15f;
+		const float k_ManipulatorSurfaceSnapBreakDist = 0.1f;
+
+		const float k_DirectSurfaceSearchScale = 1.1f;
+		const float k_DirectSurfaceSnapBreakDist = 0.03f;
+		const float k_DirectGroundSnapMin = 0.03f;
+		const float k_DirectGroundSnapMax = 0.07f;
+
+		const float k_WidgetScale = 0.03f;
+
+		const string k_SnappingEnabled = "EditorVR.Snapping.Enabled";
+		const string k_GroundSnapping = "EditorVR.Snapping.Ground";
+		const string k_SurfaceSnapping = "EditorVR.Snapping.Sufrace";
+		const string k_PivotSnapping = "EditorVR.Snapping.Pivot";
+		const string k_SnapRotation = "EditorVR.Snapping.Rotation";
+		const string k_LocalOnly = "EditorVR.Snapping.LocalOnly";
+		const string k_ManipulatorSnapping = "EditorVR.Snapping.Manipulator";
+		const string k_DirectSnapping = "EditorVR.Snapping.Direct";
+
+		[SerializeField]
+		GameObject m_GroundPlane;
+
+		[SerializeField]
+		GameObject m_Widget;
+
+		[SerializeField]
+		GameObject m_SettingsMenuPrefab;
+
+		SnappingModuleUI m_SnappingModuleUI;
+
+		class SnappingState
+		{
+			public Vector3 currentPosition;
```

Comment:
Public fields aren't allowed, so either autoprop these or turn this into a struct.